### ISSUE TITLE
Unable to mock Router with null-safety mockito

### ISSUE THREAD

**Issue by GZGavinZhao** [angulardart-community/angular#23]
### Which ng* package(s) are the source of the bug?

ngrouter

### Which operating system(s) does this bug appear on?

Linux

### Which browser(s) does this bug appear on?

N/A

### Is this a regression?

No

### Description

When using the null-safety `mockito` (^5.0.0) to mock a `Router` (with the null-safety angular of course), `testBed.create()` fails when calling the `stream` getter of the `Router` class. Before in the pre-null-safety era, `stream` would just return `null`, and since no code seems to call it, we could get away with this. Now with null safety, you cannot escape it. You either get a missing stub error or `Stream<RouterState>` is not `null` error.

It's likely that we have to put a code directly after `setup`, but that would be an extra boilerplate. It would be best if we can avoid this.

### Please provide the steps to reproduce the bug

Reproducible repo: https://github.com/GZGavinZhao/mockbug

The repo showcases the original version (how we would write it in pre-5.0 `mockito`, and three attempts to overcome this bug, both of which fail.

1. Original version:
```dart
import 'package:mockbug/app_component.template.dart' as ng;

import 'app_po.dart';

@GenerateMocks([Router])
import 'router_1_test.mocks.dart';

import 'router_1_test.template.dart' as self;

late NgTestFixture<AppComponent> fixture;
late AppPo po;

@GenerateInjector([
  ValueProvider.forToken(appBaseHref, '/'),
  routerProviders,
  ClassProvider(Router, useClass: MockRouter),
])
final InjectorFactory rootInjector = self.rootInjector$Injector;

void main() {
  final injector = InjectorProbe(rootInjector);
  final testBed = NgTestBed<AppComponent>(
    ng.AppComponentNgFactory,
    rootInjector: injector.factory,
  );

  setUp(() async {
    fixture = await testBed.create();
    final context =
        HtmlPageLoaderElement.createFromElement(fixture.rootElement);
    po = AppPo.create(context);
  });

  test('the default title should be foo', () async {
    expect(po.title, 'Foo');
  });

  test('router is not called', () async {});

  tearDown(disposeAnyRunningTest);
}
```
2. Attempt 1. We only list the part that changed.
```dart
@GenerateMocks(
  [],
  customMocks: [
    MockSpec<RouterImpl>(
      as: #MockRouter,
      returnNullOnMissingStub: true,
    ),
  ],
)
import 'router_2_test.mocks.dart';
```
3. Attempt 2.
```dart
@GenerateMocks(
  [],
  customMocks: [
    MockSpec<RouterImpl>(
      as: #MockRouter,
      returnNullOnMissingStub: true,
      fallbackGenerators: {
        #stream: mockStream,
      },
    ),
  ],
)
import 'router_3_test.mocks.dart';

import 'router_3_test.template.dart' as self;

Stream<RouterState> mockStream() {
  return Stream.empty();
}
```
4. Attempt 3.
```dart
@GenerateMocks([Router])
import 'router_4_test.mocks.dart';

void main() {
  setUp(() async {
    final mockRouter = injector.get<MockRouter>(Router);
    when(mockRouter.stream).thenAnswer((realInvocation) => Stream.empty());

    fixture = await testBed.create();
    final context =
        HtmlPageLoaderElement.createFromElement(fixture.rootElement);
    po = AppPo.create(context);
  });

  // tests...
}
```

`InjectorProbe` is a shared code here:
```dart
class InjectorProbe {
  InjectorFactory _parent;
  Injector? _injector;

  InjectorProbe(this._parent);

  InjectorFactory get factory => _factory;
  Injector? get injector => _injector;

  // Injector _factory([Injector? parent]) => _injector = _parent(parent);
  Injector _factory(Injector parent) => _injector = _parent(parent);
  T get<T>(dynamic token) => injector?.get(token);
}
```

### Please provide the exception or error you saw

```true
Original: 
  MissingStubError: 'stream'
  No stub was found which matches the arguments of this method call:
  stream
  
  Add a stub for this method using Mockito's 'when' API, or generate the MockRouter mock with a MockSpec with 'returnNullOnMissingStub: true' (see https://pub.dev/documentation/mockito/latest/annotations/MockSpec-class.html).
  dart:sdk_internal                                                        throw_
  package:mockito/src/mock.dart 191:7                                      [_noSuchMethod]
  package:mockito/src/mock.dart 185:45                                     noSuchMethod
  router_1_test.mocks.dart 47:14                                           get stream
  package:ngrouter/src/directives/router_link_active_directive.dart 44:36  ngAfterViewInit
  package:mockbug/app_component.template.dart 165:36                       detectChangesInternal
  package:ngdart/src/core/linker/views/component_view.dart 158:7           detectChangesDeprecated
  package:ngdart/src/core/linker/views/host_view.dart 182:19               detectChangesInternal
  package:ngdart/src/core/linker/views/host_view.dart 147:7                detectChangesDeprecated
  package:ngdart/src/core/change_detection/change_detector_ref.dart 175:5  DeprecatedDetectChanges.detectChanges
  package:ngdart/src/core/change_detection/host.dart 102:20                [_runTick]
  package:ngdart/src/core/change_detection/host.dart 78:7                  tick
  package:ngdart/src/core/application_ref.dart 113:5                       [_loadedRootComponent]
  package:ngdart/src/core/application_ref.dart 170:5                       TestApplicationRef.registerRootComponent
  package:ngtest/src/bootstrap.dart 93:12                                  loadComponent
  package:ngtest/src/bootstrap.dart 105:12                                 _runAndLoadComponent
  package:ngtest/src/bootstrap.dart 58:7                                   <fn>
  package:ngdart/src/core/change_detection/host.dart 207:26                <fn>
  package:ngdart/src/core/zone/ng_zone.dart 105:18                         <fn>
  dart:sdk_internal                                                        run
  package:ngdart/src/core/zone/ng_zone.dart 102:18                         [_run]
  dart:sdk_internal                                                        run
  package:ngdart/src/core/zone/ng_zone.dart 273:23                         run
  package:ngdart/src/core/application_ref.dart 148:52                      runInZone
  package:ngdart/src/core/change_detection/host.dart 205:5                 run
  package:ngtest/src/bootstrap.dart 51:16                                  bootstrapForTest

2/3:
  Expected a value of type 'Stream<RouterState>', but got one of type 'Null'
  dart:sdk_internal                                                        as_C
  router_3_test.mocks.dart 42:42                                           get stream
  package:ngrouter/src/directives/router_link_active_directive.dart 44:36  ngAfterViewInit
  package:mockbug/app_component.template.dart 165:36                       detectChangesInternal
  package:ngdart/src/core/linker/views/component_view.dart 158:7           detectChangesDeprecated
  package:ngdart/src/core/linker/views/host_view.dart 182:19               detectChangesInternal
  package:ngdart/src/core/linker/views/host_view.dart 147:7                detectChangesDeprecated
  package:ngdart/src/core/change_detection/change_detector_ref.dart 175:5  DeprecatedDetectChanges.detectChanges
  package:ngdart/src/core/change_detection/host.dart 102:20                [_runTick]
  package:ngdart/src/core/change_detection/host.dart 78:7                  tick
  package:ngdart/src/core/application_ref.dart 113:5                       [_loadedRootComponent]
  package:ngdart/src/core/application_ref.dart 170:5                       TestApplicationRef.registerRootComponent
  package:ngtest/src/bootstrap.dart 93:12                                  loadComponent
  package:ngtest/src/bootstrap.dart 105:12                                 _runAndLoadComponent
  package:ngtest/src/bootstrap.dart 58:7                                   <fn>
  package:ngdart/src/core/change_detection/host.dart 207:26                <fn>
  package:ngdart/src/core/zone/ng_zone.dart 105:18                         <fn>
  dart:sdk_internal                                                        run
  package:ngdart/src/core/zone/ng_zone.dart 102:18                         [_run]
  dart:sdk_internal                                                        run
  package:ngdart/src/core/zone/ng_zone.dart 273:23                         run
  package:ngdart/src/core/application_ref.dart 148:52                      runInZone
  package:ngdart/src/core/change_detection/host.dart 205:5                 run
  package:ngtest/src/bootstrap.dart 51:16                                  bootstrapForTest

4:
  Expected a value of type 'MockRouter', but got one of type 'Null'
  dart:sdk_internal         as_C
  utils.dart 14:30          get
  router_4_test.dart 38:29  <fn>
  dart:sdk_internal         _async
  router_4_test.dart 37:9   <fn>
```

### Please provide the dependency environment you discovered this bug in (run `dart pub deps -s compact`)

```true
Dart SDK 2.17.6
Flutter SDK 3.0.5
mockbug 0.0.1

dependencies:
- ngdart 7.1.0 [built_collection built_value collection intl js logging meta stack_trace stream_transform analyzer ngast ngcompiler build build_config code_builder csslib path lints quiver source_gen source_span]
- ngrouter 3.1.0 [ngdart collection js meta]

dev dependencies:
- build_runner 2.1.11 [args async analyzer build build_config build_daemon build_resolvers build_runner_core code_builder collection crypto dart_style frontend_server_client glob graphs http_multi_server io js logging meta mime package_config path pool pub_semver pubspec_parse shelf shelf_web_socket stack_trace stream_transform timing watcher web_socket_channel yaml]
- build_test 2.1.5 [async build build_config build_resolvers crypto glob html logging matcher package_config path stream_transform test test_core watcher]
- build_version 2.1.1 [build build_config pubspec_parse]
- build_web_compilers 3.2.3 [analyzer archive bazel_worker build build_config build_modules collection glob js logging path pool scratch_space source_maps source_span stack_trace]
- lints 2.0.0
- meta 1.8.0
- mockito 5.2.0 [analyzer build code_builder collection dart_style matcher meta path source_gen test_api]
- ngpageloader 5.0.0-dev.2 [analyzer build built_value build_config js matcher quiver source_gen webdriver collection]
- ngtest 4.1.0 [ngdart collection meta pedantic]
- test 1.21.4 [analyzer async boolean_selector collection coverage http_multi_server io js node_preamble package_config path pool shelf shelf_packages_handler shelf_static shelf_web_socket source_span stack_trace stream_channel typed_data web_socket_channel webkit_inspection_protocol yaml test_api test_core]

transitive dependencies:
- _fe_analyzer_shared 41.0.0 [meta]
- analyzer 4.2.0 [_fe_analyzer_shared collection convert crypto glob meta package_config path pub_semver source_span watcher yaml]
- archive 3.3.0 [crypto path]
- args 2.3.1
- async 2.9.0 [collection meta]
- bazel_worker 1.0.1 [async protobuf]
- boolean_selector 2.1.0 [source_span string_scanner]
- build 2.3.0 [analyzer async convert crypto glob logging meta path]
- build_config 1.0.0 [checked_yaml json_annotation path pubspec_parse yaml]
- build_daemon 3.1.0 [built_collection built_value http_multi_server logging path pool shelf shelf_web_socket stream_transform watcher web_socket_channel]
- build_modules 4.0.5 [analyzer async bazel_worker build build_config collection crypto glob graphs json_annotation logging path scratch_space stream_transform]
- build_resolvers 2.0.9 [analyzer async build crypto graphs logging path package_config pool pub_semver stream_transform yaml]
- build_runner_core 7.2.3 [async build build_config build_resolvers collection convert crypto glob graphs json_annotation logging meta path package_config pool timing watcher yaml]
- built_collection 5.1.1
- built_value 8.3.3 [built_collection collection fixnum meta]
- charcode 1.3.1
- checked_yaml 2.0.1 [json_annotation source_span yaml]
- clock 1.1.1
- code_builder 4.1.0 [built_collection built_value collection matcher meta]
- collection 1.16.0
- convert 3.0.2 [typed_data]
- coverage 1.3.2 [args logging package_config path source_maps stack_trace vm_service]
- crypto 3.0.2 [typed_data]
- csslib 0.17.2 [source_span]
- dart_style 2.2.3 [analyzer args path pub_semver source_span]
- file 6.1.2 [meta path]
- fixnum 1.0.1
- frontend_server_client 2.1.3 [async path]
- glob 2.1.0 [async collection file path string_scanner]
- graphs 2.1.0 [collection]
- html 0.15.0 [csslib source_span]
- http_multi_server 3.2.1 [async]
- http_parser 4.0.1 [collection source_span string_scanner typed_data]
- intl 0.17.0 [clock path]
- io 1.0.3 [meta path string_scanner]
- js 0.6.4
- json_annotation 4.6.0 [meta]
- logging 1.0.2
- matcher 0.12.11 [stack_trace]
- mime 1.0.2
- ngast 2.1.3 [analyzer charcode collection meta source_span string_scanner]
- ngcompiler 2.1.1+0 [analyzer ngdart ngast args build build_test code_builder collection csslib dart_style logging meta package_config path source_gen source_span stack_trace]
- node_preamble 2.0.1
- package_config 2.1.0 [path]
- path 1.8.2
- pedantic 1.11.1
- pool 1.5.1 [async stack_trace]
- protobuf 2.1.0 [fixnum collection meta]
- pub_semver 2.1.1 [collection meta]
- pubspec_parse 1.2.0 [checked_yaml collection json_annotation pub_semver yaml]
- quiver 3.1.0 [matcher]
- scratch_space 1.0.1 [build crypto path pool]
- shelf 1.3.1 [async collection http_parser path stack_trace stream_channel]
- shelf_packages_handler 3.0.1 [path shelf shelf_static]
- shelf_static 1.1.1 [convert http_parser mime path shelf]
- shelf_web_socket 1.0.2 [shelf stream_channel web_socket_channel]
- source_gen 1.2.2 [analyzer async build dart_style glob meta path source_span yaml]
- source_map_stack_trace 2.1.0 [path stack_trace source_maps]
- source_maps 0.10.10 [source_span]
- source_span 1.9.0 [collection path term_glyph]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- stream_transform 2.0.0
- string_scanner 1.1.1 [source_span]
- sync_http 0.3.1
- term_glyph 1.2.1
- test_api 0.4.12 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph matcher]
- test_core 0.4.16 [analyzer async args boolean_selector collection coverage frontend_server_client glob io meta package_config path pool source_map_stack_trace source_maps source_span stack_trace stream_channel vm_service yaml matcher test_api]
- timing 1.0.0 [json_annotation]
- typed_data 1.3.1 [collection]
- vm_service 8.3.0
- watcher 1.0.1 [async path]
- web_socket_channel 2.2.0 [async crypto stream_channel]
- webdriver 3.0.0 [archive matcher path stack_trace sync_http]
- webkit_inspection_protocol 1.1.0 [logging]
- yaml 3.1.1 [collection source_span string_scanner]
```

**Comment by denniskaselow** [comment]
There are two options I am aware of.

First, without code generation, create your own custom mock, but you'll either have to stub or implement all methods that are called:

```dart
class MockRouter extends Mock implements Router {
  @override
  Stream<RouterState> get stream => Stream.empty();
  @override
  RouterState? get current => null;
}
```

Second, based on your second variant, but with a `FactoryProvider` and code added to `setUp`:

```dart
late Router mockRouter;

Router mockRouterProvider() => mockRouter;

@GenerateInjector([
  ValueProvider.forToken(appBaseHref, '/'),
  routerProviders,
  FactoryProvider(Router, mockRouterProvider),
])
final InjectorFactory rootInjector = self.rootInjector$Injector;
```

but then you'll have to stub the call in `setUp`:

```dart
  setUp(() async {
    mockRouter = MockRouter();
    when(mockRouter.stream).thenAnswer((_) => Stream.empty());
    fixture = await testBed.create();
    final context =
        HtmlPageLoaderElement.createFromElement(fixture.rootElement);
    po = AppPo.create(context);
  });
```

Using you the original variant, you'd also need to stub the call to `current`.

No idea what exactly `fallbackGenerators` is used for, but that won't work:

> A fallback value is not ever exposed in stubbing or verifying; it is an object that mockito's internals can use as a legal return value.

Opening an issue for something that actually is exposed for stubbing is another solution. My issues were handled swiftly when mockito moved to null safety and I ran into some bugs.

**Comment by denniskaselow** [comment]
@GZGavinZhao have you seen my comment? Is it a viable option?

**Comment by GZGavinZhao** [comment]
@denniskaselow Sorry for the late response. It does compile, but I'm thinking about how to efficiently write [actual tests](https://angulardart.xyz/guide/testing/component/routing-components) using the `factoryProvider`.

**Comment by GZGavinZhao** [comment]
@denniskaselow Haven't found a better method, so I'll use your method for now to prevent this issue from becoming a blocker. Thank you for your suggestions!

**Comment by GZGavinZhao** [comment]
Update: that was some really good timing. `mockito` just released (`v5.3.0`) something called `GenerateNiceMocks`, which will generate mocks that try to return some default values when an unstubbed method is called.
```dart
@GenerateNiceMocks([MockSpec<Router>(as: #MockRouter)])
```
I think this is probably the easiest way to go.
